src/atoms/userSettings.ts

```diff
@@ -1,23 +1,108 @@
+// const editorModeAtom = atom<EditorMode>('Normal');
+// editorModeAtom.onMount = setValue => {
+//   // this might trigger hydration issues? whatever
+//   if (typeof window !== 'undefined') {
+//     const editorMode = window.localStorage.getItem('editorMode');
+//     if (editorMode === 'Vim' || editorMode === 'Normal') {
+//       setValue(editorMode);
+//     }
+//   }
+// };
+// export const editorModeAtomWithPersistence = atom<EditorMode, EditorMode>(
+//   get => get(editorModeAtom),
+//   (_get, set, val) => {
+//     set(editorModeAtom, val);
+//     localStorage.setItem('editorMode', val);
+//   }
+// );
+
+export const userSettingsRefAtom = atom<firebaseType.database.Reference | null>(
+  get => {
+    const firebaseUser = get(firebaseUserAtom);
+    if (!firebaseUser) {
+      return null;
     }
+    const uid = firebaseUser.uid;
+    return firebase.database().ref('users').child(uid).child('settings');
   }
+);
+
+interface UserSettings {
+  name: string;
+  color: string;
+  editorMode: EditorMode;
+  defaultPermission: DefaultPermission;
+  defaultLang: Language;
+}
+
+const defaultUserSettings: UserSettings = {
+  name: '', // change in settings
+  editorMode: 'Normal', // change in settings
+  color: '', // fixed
+  defaultPermission: 'READ_WRITE', // change in dashboard
+  defaultLang: 'cpp', // last viewed file
 };
-export const editorModeAtomWithPersistence = atom<EditorMode, EditorMode>(
-  get => get(editorModeAtom),
-  (get, set, val) => {
-    set(editorModeAtom, val);
-    localStorage.setItem('editorMode', val);
+
+export const baseDisplayNameAtom = atom<string>('');
+export const displayNameAtom = atom(
+  get => get(baseDisplayNameAtom),
+  (get, set, displayName: string) => {
+    const firebaseUser = get(firebaseUserAtom);
+    if (firebaseUser) firebaseUser.updateProfile({ displayName });
+    set(baseDisplayNameAtom, displayName);
   }
 );
 
-export const userNameAtom = atom<string | null>(null);
+export const baseUserSettingsAtom = atom<UserSettings>(defaultUserSettings);
+export const _userSettingsAtom = atom<UserSettings, Partial<UserSettings>>(
+  get => {
+    const firebaseUser = get(firebaseUserAtom);
+    let obj = { ...get(baseUserSettingsAtom) };
+    if (firebaseUser) {
+      obj = {
+        ...obj,
+        name: get(displayNameAtom),
+        color: colorFromUserId(firebaseUser.uid),
+      };
+    }
+    return obj;
+  },
+  (get, set, val) => {
+    // only use this in WorkspaceInitializer
+    set(baseUserSettingsAtom, {
+      ...get(baseUserSettingsAtom),
+      ...val,
+    });
+  }
+);
+export const userSettingsAtomWithPersistence = atom<
+  UserSettings,
+  Partial<UserSettings>
+>(
+  get => get(_userSettingsAtom),
+  (get, _set, val) => {
+    const data: UserSettings = {
+      ...get(_userSettingsAtom),
+      ...val,
+    };
+    const userRef = get(userSettingsRefAtom);
+    if (userRef) userRef.update(data);
+    else {
+      // fail silently, it's probably okay?
```

Comment:
oops this is questionable. Else statement should only be triggered when `val = { defaultLang }` and the URL contains a `queryParam`. In this case, the last viewed lang of the user is not correctly updated.

Specifically, consider the following sequence of steps:

 1. You view a file in Java. 
 2. You open a link that ends with `?lang=py`. 
 3. You create a new workspace in a new tab.

Then the workspace will open to Java, not Python (though Python would make more sense). What's the best way to fix this?

Reply:
One way is to only allow this to be called after firebase is initialized. I think it's fine as is right now though...